File: backend/app/api/v1/budgets.py
import uuid
from datetime import date
from fastapi import APIRouter, Query, status

from app.api.deps import current_user_dep, db_dep
from app.schemas.budget import BudgetCreate, BudgetResponse, BudgetStatusResponse, BudgetUpdate
from app.services import budget_service

router = APIRouter()


@router.get("", response_model=list[BudgetResponse])
async def list_budgets(
    db: db_dep,
    current_user: current_user_dep,
    period_month: date | None = Query(
        None, description="Month of the budgets (YYYY-MM-DD)"
    ),
) -> list[BudgetResponse]:
    """List budgets configured for authenticated user."""
    return await budget_service.get_budgets(db, current_user.id, period_month)


@router.post("", response_model=BudgetResponse, status_code=status.HTTP_201_CREATED)
async def create_budget(
    db: db_dep,
    current_user: current_user_dep,
    payload: BudgetCreate,
) -> BudgetResponse:
    """Create a new budget goal (overall or per-category) for authenticated user."""
    return await budget_service.create_budget(db, current_user.id, payload)


@router.get("/status", response_model=list[BudgetStatusResponse])
async def get_budgets_status(
    db: db_dep,
    current_user: current_user_dep,
    period_month: date | None = Query(
        None, description="Month for checking status (defaults to current month)"
    ),
) -> list[BudgetStatusResponse]:
    """Retrieve live warning status, spent, and remaining balance for authenticated user's budgets."""
    if period_month is None:
        period_month = date.today()
    return await budget_service.get_budgets_status(db, current_user.id, period_month)


@router.get("/{budget_id}", response_model=BudgetResponse)
async def get_budget(
    db: db_dep,
    current_user: current_user_dep,
    budget_id: uuid.UUID,
) -> BudgetResponse:
    """Get details of a single user-owned budget goal."""
    return await budget_service.get_budget_by_id(db, current_user.id, budget_id)


@router.put("/{budget_id}", response_model=BudgetResponse)
async def update_budget(
    db: db_dep,
    current_user: current_user_dep,
    budget_id: uuid.UUID,
    payload: BudgetUpdate,
) -> BudgetResponse:
    """Update limit amount of a user-owned budget goal."""
    return await budget_service.update_budget(db, current_user.id, budget_id, payload)


@router.delete("/{budget_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_budget(
    db: db_dep,
    current_user: current_user_dep,
    budget_id: uuid.UUID,
) -> None:
    """Remove a user-owned budget goal."""
    await budget_service.delete_budget(db, current_user.id, budget_id)

File: frontend/src/lib/queries/budgets.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { apiClient } from "../api-client";
import { Budget, BudgetStatus } from "../../types/budget";

export const BUDGETS_QUERY_KEY = ["budgets"];

// Fetch list of budgets configured for a period
export function useBudgets(periodMonth?: string) {
  return useQuery<Budget[], any>({
    queryKey: [...BUDGETS_QUERY_KEY, periodMonth],
    queryFn: async () => {
      const params = periodMonth ? { period_month: periodMonth } : undefined;
      const response = await apiClient.get<Budget[]>("/budgets", { params });
      return response.data;
    },
  });
}

// Fetch live warning status, spent, and remaining details for all budgets
export function useBudgetsStatus(periodMonth?: string) {
  return useQuery<BudgetStatus[], any>({
    queryKey: [...BUDGETS_QUERY_KEY, "status", periodMonth],
    queryFn: async () => {
      const params = periodMonth ? { period_month: periodMonth } : undefined;
      const response = await apiClient.get<BudgetStatus[]>("/budgets/status", { params });
      return response.data;
    },
  });
}

// Create overall or per-category budget goal
export function useCreateBudget() {
  const queryClient = useQueryClient();
  
  return useMutation<Budget, any, Omit<Budget, "id" | "category" | "created_at" | "updated_at">>({
    mutationFn: async (payload) => {
      const response = await apiClient.post<Budget>("/budgets", payload);
      return response.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: BUDGETS_QUERY_KEY });
      queryClient.invalidateQueries({ queryKey: ["dashboard"] });
    },
  });
}

// Update limit amount of a budget goal
export function useUpdateBudget() {
  const queryClient = useQueryClient();
  
  return useMutation<Budget, any, { id: string; limit_amount: number }>({
    mutationFn: async ({ id, limit_amount }) => {
      const response = await apiClient.put<Budget>(`/budgets/${id}`, { limit_amount });
      return response.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: BUDGETS_QUERY_KEY });
      queryClient.invalidateQueries({ queryKey: ["dashboard"] });
    },
  });
}

// Remove a budget goal
export function useDeleteBudget() {
  const queryClient = useQueryClient();
  
  return useMutation<void, any, string>({
    mutationFn: async (id) => {
      await apiClient.delete(`/budgets/${id}`);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: BUDGETS_QUERY_KEY });
      queryClient.invalidateQueries({ queryKey: ["dashboard"] });
    },
  });
}
